Mathlib/Data/Set/Pointwise/Basic.lean

```diff
@@ -0,0 +1,1189 @@
+  { Set.involutiveNeg with
+    neg_mul := fun _ _ => by
+      simp_rw [← image_neg]
+      exact image2_image_left_comm neg_mul
+    mul_neg := fun _ _ => by
+      simp_rw [← image_neg]
+      exact image_image2_right_comm mul_neg }
+#align set.has_distrib_neg Set.hasDistribNeg
+
+scoped[Pointwise]
+  attribute [instance]
+    Set.divisionMonoid Set.subtractionMonoid Set.divisionCommMonoid Set.subtractionCommMonoid
+    Set.hasDistribNeg
+
+section Distrib
+
+variable [Distrib α] (s t u : Set α)
+
+/-!
+Note that `Set α` is not a `Distrib` because `s * t + s * u` has cross terms that `s * (t + u)`
+lacks.
+-/
+
+
+theorem mul_add_subset : s * (t + u) ⊆ s * t + s * u :=
+  image2_distrib_subset_left mul_add
+#align set.mul_add_subset Set.mul_add_subset
+
+theorem add_mul_subset : (s + t) * u ⊆ s * u + t * u :=
+  image2_distrib_subset_right add_mul
+#align set.add_mul_subset Set.add_mul_subset
+
+end Distrib
+
+section MulZeroClass
+
+variable [MulZeroClass α] {s t : Set α}
+
+/-! Note that `Set` is not a `MulZeroClass` because `0 * ∅ ≠ 0`. -/
+
+
+theorem mul_zero_subset (s : Set α) : s * 0 ⊆ 0 := by simp [subset_def, mem_mul]
+#align set.mul_zero_subset Set.mul_zero_subset
+
+theorem zero_mul_subset (s : Set α) : 0 * s ⊆ 0 := by simp [subset_def, mem_mul]
+#align set.zero_mul_subset Set.zero_mul_subset
+
+theorem Nonempty.mul_zero (hs : s.Nonempty) : s * 0 = 0 :=
+  s.mul_zero_subset.antisymm <| by simpa [mem_mul] using hs
+#align set.nonempty.mul_zero Set.Nonempty.mul_zero
+
+theorem Nonempty.zero_mul (hs : s.Nonempty) : 0 * s = 0 :=
+  s.zero_mul_subset.antisymm <| by simpa [mem_mul] using hs
+#align set.nonempty.zero_mul Set.Nonempty.zero_mul
+
+end MulZeroClass
+
+section Group
+
+variable [Group α] {s t : Set α} {a b : α}
+
+/-! Note that `Set` is not a `Group` because `s / s ≠ 1` in general. -/
+
+
+@[to_additive (attr := simp)]
+theorem one_mem_div_iff : (1 : α) ∈ s / t ↔ ¬Disjoint s t := by
+  simp [not_disjoint_iff_nonempty_inter, mem_div, div_eq_one, Set.Nonempty]
+#align set.one_mem_div_iff Set.one_mem_div_iff
+
+@[to_additive]
+theorem not_one_mem_div_iff : (1 : α) ∉ s / t ↔ Disjoint s t :=
+  one_mem_div_iff.not_left
+#align set.not_one_mem_div_iff Set.not_one_mem_div_iff
+
+alias not_one_mem_div_iff ↔ _ _root_.Disjoint.one_not_mem_div_set
+
+attribute [to_additive] Disjoint.one_not_mem_div_set
+
+@[to_additive]
+theorem Nonempty.one_mem_div (h : s.Nonempty) : (1 : α) ∈ s / s :=
+  let ⟨a, ha⟩ := h
+  mem_div.2 ⟨a, a, ha, ha, div_self' _⟩
+#align set.nonempty.one_mem_div Set.Nonempty.one_mem_div
+
+@[to_additive]
+theorem isUnit_singleton (a : α) : IsUnit ({a} : Set α) :=
+  (Group.isUnit a).set
+#align set.is_unit_singleton Set.isUnit_singleton
+
+@[to_additive (attr := simp)]
+theorem isUnit_iff_singleton : IsUnit s ↔ ∃ a, s = {a} := by
+  simp only [isUnit_iff, Group.isUnit, and_true_iff]
+#align set.is_unit_iff_singleton Set.isUnit_iff_singleton
```

Comment:
```lean
#check Set.isAddUnit_iff_singleton.{u_1} /- Left-hand side simplifies from
  IsAddUnit s
to
  ∃ a, s = {a} ∧ IsAddUnit a
using
  simp only [@Set.isAddUnit_iff]
Try to change the left-hand side to the simplified term!
 -/
#check Set.isUnit_iff_singleton.{u_1} /- Left-hand side simplifies from
  IsUnit s
to
  ∃ a, s = {a} ∧ IsUnit a
using
  simp only [@Set.isUnit_iff]
Try to change the left-hand side to the simplified term!
 -/
```

Not sure what to do here. Does it need priority? How does that work with `to_additive`?